### DISCUSSION TITLE
Update InsertNode on callback

### DISCUSSION THREAD

**Discussion by tiagovla** [L3MON4D3/LuaSnip#449] [Q&A] [answered]
I was trying to use wolframscript inside a snippet.

Is there a way to use set_text on an InsertNode and update the text inside the callback using the api?

So far, I got this working:
```lua
ls.add_snippets("tex", {
    s({ trig = "math" }, { i(1), t " ", i(0) }, {
        callbacks = {
            [1] = {
                [events.leave] = function(node)
                    local from_pos, to_pos = node.mark:pos_begin_end_raw()
                    local lines = vim.api.nvim_buf_get_lines(0, from_pos[1], to_pos[1] + 1, false)
                    local script = ([[!wolframscript -c "ToString@TeXForm[%s]"]]):format(node:get_text()[1])
                    local output = vim.split(vim.api.nvim_exec(script, true), "\n")
                    output = output[#output - 1]
                    if #lines == 1 then
                        local p1 = string.sub(lines[1], 1, from_pos[2])
                        local p2 = string.sub(lines[1], to_pos[2] + 1, #lines[1])
                        lines[1] = p1 .. output .. p2
                    end
                    vim.api.nvim_buf_set_lines(0, from_pos[1], to_pos[1] + 1, false, lines)
                    vim.api.nvim_win_set_cursor(0, { to_pos[1] + 1, from_pos[2] + #output }) -- nope
                end,
            },
        },
    }),
}, {})
```

**Comment by L3MON4D3** [comment, marked as answer]
I think we could support a `set_text` for insertNode and textNode, it would probably look very much like what you have currently.
Is the `win_set_cursor` necessary? Afaict the cursor should be set after jumping.

**Reply by tiagovla** [reply]
I was trying to figure it out why the cursor was being moved to the begging of the line:
![gifcast_220601083701](https://user-images.githubusercontent.com/30515389/171395943-e5fe1308-4da8-4afe-a330-a6629e4d98ca.gif)
That line with `win_set_cursor` is ignored. Maybe I should update the node's marks?

**Reply by L3MON4D3** [reply]
Ah ofc, I didn't spot the set_lines, could you try with a `nvim_buf_set_text` with exactly the node's range?
(node-positions are tracked with extmarks, set_lines moves all of them to the beginning or end, while set_text will leave those outside the texts range alone, even on the same line)

**Reply by tiagovla** [reply]
> Ah ofc, I didn't spot the set_lines, could you try with a `nvim_buf_set_text` with exactly the node's range? (node-positions are tracked with extmarks, set_lines moves all of them to the beginning or end, while set_text will leave those outside the texts range alone, even on the same line)

True! I forgot about that. It does work as it should now. Thanks!
